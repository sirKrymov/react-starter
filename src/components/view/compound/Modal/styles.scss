@import '../../../../styles/configs';
@import '../../../../styles/mixins';

.modal {
  pointer-events: none;
  overflow: hidden;
  position: fixed;
  z-index: 1000000;
  bottom: 0;
  right: 0;
  left: 0;
  top: 0;
  @include center;

  &--open {
    pointer-events: auto;
    z-index: 1000001;

    .modal__content {
      animation: modal-content 0.4s 0.2s forwards;
      transform: translateY(40px);
      opacity: 0;
    }

    .modal__backdrop {
      animation: show 0.15s forwards;
      opacity: 0;
    }
  }

  &__backdrop {
    background-color: rgba(255, 255, 255, 0.4);
    backdrop-filter: blur(2px);
    will-change: opacity;
    position: absolute;
    bottom: 0;
    right: 0;
    left: 0;
    top: 0;
  }

  &__content {
    background-color: var(--primary);
    border-radius: $border-radius;
    will-change: transform, opacity;
    box-shadow: $box-shadow-modal;
    position: relative;
    min-height: 240px;
    overflow: auto;
    z-index: 100;
    margin: 1.5rem;
    width: 100%;

    &--width-xs {
      max-width: 460px;
    }

    &--width-sm {
      max-width: 575px;
    }

    &--width-md {
      max-width: 767px;
    }

    &--width-min-content {
      max-width: initial;
      width: initial;
    }

    &--width-full {
      max-width: 100%;
    }

    @media (max-width: $breakpoint-down-xs) {
      border-radius: 0;
      max-width: 100%;
      height: 100vh;
      margin: 0;
    }
  }

  &__body {
    padding: 0 1.5rem 1.5rem;

    &--no-padding {
      padding: 0;
    }
  }
}
