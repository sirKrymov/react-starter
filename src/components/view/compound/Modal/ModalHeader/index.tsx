// @SETUP: Please don't forget to remove this component or current comment if component is used in project
import React, { ReactElement } from 'react';

import { IoMdCloseCircle } from 'react-icons/io';

import { Typography } from '../../../atoms/Typography';
import { ReactIcon } from '../../../atoms/ReactIcon';

import './styles.scss';

interface IProps {
  onClose?(): void;
  title?: string;
}

export function ModalHeader({ onClose, title }: IProps): ReactElement | null {
  return onClose || title ? (
    <div className="modal-header">
      {title && (
        <Typography className="modal-header__title" variant="h3" noWrap>
          {title}
        </Typography>
      )}

      {onClose && (
        <ReactIcon
          className="modal-header__close-btn"
          onClick={onClose}
          size="xl"
        >
          <IoMdCloseCircle />
        </ReactIcon>
      )}
    </div>
  ) : null;
}
