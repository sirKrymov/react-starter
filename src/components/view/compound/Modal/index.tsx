// @SETUP: Please don't forget to remove this component or current comment if component is used in project
import React, { useEffect, ReactNode, ReactElement } from 'react';

import { createPortal } from 'react-dom';
import cn from 'classnames';

import { ModalHeader } from './ModalHeader';

import './styles.scss';

interface IProps {
  bodyHeight?: 'default' | 'full';
  noPadding?: boolean;
  closeIcon?: boolean;
  onClose?(): void;
  children: ReactNode;
  width?: 'xs' | 'sm' | 'md' | 'min-content' | 'full';
  title?: string;
  open: boolean;
}

export function Modal({
  bodyHeight,
  noPadding,
  closeIcon = true,
  children,
  width = 'sm',
  onClose,
  title,
  open
}: IProps): ReactElement {
  useEffect(() => {
    if (open) {
      setTimeout(() => {
        document.body.setAttribute('class', 'overflow-hidden');
      }, 250);
    } else {
      document.body.removeAttribute('class');
    }
  }, [open]);

  const modalClass = cn({
    modal: true,
    'modal--open': open
  });
  const modalContent = cn({
    [`modal__content`]: true,
    [`modal__content--width-${width}`]: width,
    [`modal__content--body-height-${bodyHeight}`]: bodyHeight
  });
  const modalBodyClass = cn({
    [`modal__body`]: true,
    ' modal__content--no-padding': noPadding
  });

  return createPortal(
    open ? (
      <div className={modalClass}>
        <div className={modalContent}>
          {(title || (closeIcon && onClose)) && (
            <ModalHeader onClose={onClose} title={title} />
          )}

          <div className={modalBodyClass}>{children}</div>
        </div>

        <div className="modal__backdrop" />
      </div>
    ) : null,
    document.body
  );
}
