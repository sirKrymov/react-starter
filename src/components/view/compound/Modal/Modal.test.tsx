import React from 'react';

import ReactDOM from 'react-dom';

import renderer, { ReactTestRenderer } from 'react-test-renderer';

import { shallow, ShallowWrapper } from 'enzyme';

import { Modal } from '../Modal';

describe('<Modal> compound component', () => {
  ReactDOM.createPortal = (node: any) => node;

  let shallowComponent: ShallowWrapper;
  let component: ReactTestRenderer;

  const createComponent = ({ open, ...rest }: Record<string, any>) => {
    component = renderer.create(
      <Modal open={open} {...rest}>
        <div id="modal-children">Check modal test content</div>
      </Modal>
    );
  };

  const createShallowComponent = ({ open, ...rest }: Record<string, any>) => {
    shallowComponent = shallow(
      <Modal open={open} {...rest}>
        <div id="modal-children">Check modal test content</div>
      </Modal>
    );
  };

  describe('WHEN the component loads', () => {
    it('THEN it should render without crashing', () => {
      createComponent({ open: true });

      expect(component.toJSON()).toBeDefined();
    });

    it('THEN it should match snapshot', () => {
      createComponent({ open: true });

      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  describe('WHEN the component have property open=true', () => {
    it('THEN it should render the children', () => {
      createShallowComponent({ open: true });

      expect(shallowComponent.find('#modal-children')).toBeDefined();
    });

    it('should set overflow hidden on the body element', () => {
      const body = document.querySelector('body');

      if (body) {
        expect(body.hasAttribute('class')).toBeFalsy();

        createComponent({ open: true });

        setTimeout(() => {
          expect(body.hasAttribute('class')).toEqual('overflow-hidden');
        }, 250);
      }
    });
  });

  describe('WHEN the component have property open=false', () => {
    it('THEN it should`t render', () => {
      createComponent({ open: false });

      expect(component.toJSON()).toBe(null);
    });

    it('should unset overflow hidden on the body element', () => {
      const body = document.querySelector('body');

      createComponent({ open: false });

      if (body) {
        expect(body.hasAttribute('class')).toBeFalsy();
      }
    });
  });
});
