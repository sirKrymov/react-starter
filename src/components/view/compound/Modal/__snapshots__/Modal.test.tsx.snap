// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal> compound component WHEN the component loads THEN it should match snapshot 1`] = `
<div
  className="modal modal--open"
>
  <div
    className="modal__content modal__content--width-sm"
  >
    <div
      className="modal__body"
    >
      <div
        id="modal-children"
      >
        Check modal test content
      </div>
    </div>
  </div>
  <div
    className="modal__backdrop"
  />
</div>
`;
